mqtt/mqtt-broker/src/configuration.rs

```diff
@@ -0,0 +1,304 @@
+            Duration::from_secs(60 * 24 * 60 * 60)
+        );
+    }
+
+    #[test]
+    fn it_overrides_defaults() {
+        let settings = BrokerConfig::from_file(Path::new("test/config_correct.json"))
+            .expect("should be able to create instance from configuration file");
+
+        assert_eq!(
+            settings.retained_messages.expiration,
+            Duration::from_secs(90 * 24 * 60 * 60)
+        );
+    }
+
+    #[test]
+    fn it_refuses_persistence_with_no_file_path() {
+        let settings = BrokerConfig::from_file(Path::new("test/config_no_file_path.json"));
+
+        assert_matches!(settings, Err(_err));
+    }
+
+    #[test]
+    fn it_type_mismatch_fails() {
+        let settings = BrokerConfig::from_file(Path::new("test/config_bad_value_type.json"));
+
+        assert_matches!(settings, Err(_err));
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct Container {
+        #[serde(deserialize_with = "humansize")]
+        size: u64,
+    }
+
+    #[test_case( "123b",  123 ; "when using bytes")]
+    #[test_case( "123kb",  123*1024 ; "when using kilobytes")]
+    #[test_case( "123mb",  123*1024*1024 ; "when using megabytes")]
+    #[test_case( "123gb",  123*1024*1024*1024 ; "when using gigabytes")]
+    fn it_deserializes_different_multipliers(input: &str, expected: u64) {
+        let container_json = json!({ "size": input }).to_string();
+
+        let container: Container = serde_json::from_str(&container_json).unwrap();
+        assert_eq!(container.size, expected);
+    }
+
+    #[test_case( "123kb",  123*1024 ; "when using all lowercase")]
+    #[test_case( "123Kb",  123*1024 ; "when using mixed case")]
+    #[test_case( "123KB",  123*1024 ; "when using all capitals")]
+    #[test_case( "  123kb",  123*1024 ; "when using leading spaces")]
+    #[test_case( "123kb  ",  123*1024 ; "when using trailing spaces")]
+    #[test_case( "123  kb",  123*1024 ; "when using separator spaces")]
+    #[test_case( "  123  kb  ",  123*1024 ; "when using spaces at multiple positions")]
+    fn it_deserializes_different_writing_modes(input: &str, expected: u64) {
+        let container_json = json!({ "size": input }).to_string();
+
+        let container: Container = serde_json::from_str(&container_json).unwrap();
+        assert_eq!(container.size, expected);
+    }
+
+    #[test_case( "123tb" ; "when using unknown unit")]
+    #[test_case( "123" ; "when missing unit")]
+    #[test_case( "12a3 kb" ; "when invalid number")]
+    fn it_fails_deserializing_unknown_unit(input: &str) {
+        let container_json = json!({ "size": input }).to_string();
+        let result = serde_json::from_str::<Container>(&container_json);
+
+        assert_matches!(result, Err(_err));
+    }
+
+    proptest! {
+        #[test]
+        fn doesnt_crash(s in "\\PC*") {
+            parse_size_anything(&s);
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn it_parses_valid_input(p in (r"\s*",             // leading spaces if any
```

Comment:
Is it possible to move generator part in its own function? now it is a bit messy

Reply:
Propably, @myagley has more ideas on how to improve prop tests